src/org/jitsi/service/neomedia/RawPacket.java

```diff
@@ -203,107 +203,202 @@ public static long getRTCPSSRC(byte[] buf, int off, int len)
-        int newPayloadOffset = payloadOffset - offset + extraBytes;
-        int newBuffLen = length + extraBytes;
-        int lengthToCopy = FIXED_HEADER_SIZE + getCsrcCount()*4;
+        addExtension(id, data, data.length);
+    }
 
-        if (extensionBit)
+    /**
+     * Adds the given buffer as a header extension of this packet
+     * according the rules specified in RFC 5285. Note that this method does
+     * not replace extensions so if you add the same buffer twice it would be
+     * added as a separate extension.
+     *
+     * This method MUST NOT be called while iterating over the extensions using
+     * {@link #getHeaderExtensions()}, or while manipulating the state of this
+     * {@link RawPacket}.
+     *
+     * @param id the ID with which to add the extension.
+     * @param data the buffer containing the extension data.
+     * @param len the length of the extension.
+     */
+    public void addExtension(byte id, byte[] data, int len)
+    {
+        if (id < 1 || id > 15 || data == null ||
+            len < 1 || len > 16 || data.length < len)
         {
-            // without copying the extension length value, will set it later
-            lengthToCopy += EXT_HEADER_SIZE - 2;
+            throw new IllegalArgumentException(
+                "id=" + id + " data.length="
+                    + (data == null ? "null" : data.length)
+                    + " len=" + len);
         }
 
-        // Warning: newBuffer may be the same byte[] as buffer. This must be
-        // taken into account while copying or writing to newBuffer below.
-        byte[] newBuffer = buffer;
-        if (newBuffer.length < newBuffLen)
+        // The byte[] of a RawPacket has the following structure:
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        // | A: unused | B: hdr + ext | C: payload | D: unused |
+        // +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+        // And the regions have the following sizes:
+        // A: this.offset
+        // B: this.getHeaderLength()
+        // C: this.getPayloadLength()
+        // D: this.buffer.length - this.length - this.offset
+        // We will try to extend the packet so that it uses A and/or D if
+        // possible, in order to avoid allocating new memory.
+
+        // We get this early, before we modify the buffer.
+        int payloadLength = getPayloadLength();
+        boolean extensionBit = getExtensionBit();
+        int extHeaderOffset = FIXED_HEADER_SIZE + 4 * getCsrcCount();
+
+
+        // This is an upper bound on the required length for the packet after
+        // the addition of the new extension. It is easier to calculate than
+        // the exact number, and is relatively close (it may be off by a few
+        // bytes due to padding)
+        int maxRequiredLength
+            = getLength()
+            + (extensionBit ? 0 : EXT_HEADER_SIZE)
+            + len + 1 /* header extension header */
+            + 3 /* padding */;
```

Comment:
Do you think this still applies? (taken from CsrcTransformEngine.java)

    // calculate extension padding		
    int padLen = 4 - (len + 1) % 4;
    if (padLen == 4) padLen = 0;